prince-of-versions/src/main/java/co/infinum/princeofversions/PrinceOfVersions.java

```diff
@@ -1,275 +1,209 @@
- * This is important because of cancel functionality. There is no way once cancelled loader became uncancelled, so to support correct
- * cancel functionality always provide new instance of loader.
- * </p>
+ *         {@link PrinceOfVersions} updater = new {@link PrinceOfVersions}(context);
+ *         {@link PrinceOfVersionsCall} call = updater.checkForUpdates("http://example.com/some/update.json", callback); // starts checking
+ * for update
+ * </pre>
  */
 public class PrinceOfVersions {
 
-    /**
-     * Factory for creating VersionVerifier instance.
-     */
-    private VersionVerifierFactory factory;
+    private static Parser createDefaultParser() {
+        return new JsonParser();
+    }
 
-    /**
-     * Repository for persisting library data.
-     */
-    private VersionRepository repository;
+    private static Storage createDefaultStorage(Context context) {
+        return new PrinceOfVersionsDefaultStorage(context);
+    }
 
-    /**
-     * SDK int provider
-     */
-    private SdkVersionProvider sdkVersionProvider;
+    private static VersionParser createDefaultVersionParser() {
+        return new PrinceOfVersionsDefaultVersionParser();
+    }
 
-    /**
-     * Creates a new instance of updater for application associated with provided context.
-     *
-     * @param context Context of associated application.
-     */
-    public PrinceOfVersions(@NonNull final Context context) {
-        this(context.getApplicationContext(), createDefaultVersionVerifierFactory(new ParserFactory() {
-            @Override
-            public VersionConfigParser newInstance() {
-                try {
-                    return new JsonVersionConfigParser(ContextHelper.getAppVersion(context.getApplicationContext()));
-                } catch (PackageManager.NameNotFoundException e) {
-                    throw new IllegalArgumentException("Current version not available.");
-                }
-            }
-        }));
+    private static ApplicationConfiguration createAppConfig(Context context) {
+        return new ApplicationConfigurationImpl(context);
     }
 
+    private Presenter presenter;
+
+    private ApplicationConfiguration appConfig;
+
     /**
-     * Creates a new instance of updater for application associated with provided context using custom parser implementation.
+     * Creates {@link PrinceOfVersions} using provided {@link Context}.
      *
-     * @param context       Context of associated application.
-     * @param parserFactory Factory for creating custom parser for parsing loaded content.
+     * @param context context which will be used for checking application version.
      */
-    public PrinceOfVersions(@NonNull final Context context, ParserFactory parserFactory) {
-        this(context.getApplicationContext(), createDefaultVersionVerifierFactory(parserFactory));
+    public PrinceOfVersions(Context context) {
```

Comment:
Since `PrinceOfVersions.Builder().build(Context);` does exactly the same thing as this constructor, do you think it would be wise for this class to be constructible only with the builder? In my opinion, that way it would be less error-prone and less confusing for developers.

Reply:
I'm fine with that proposal. However I don't have anything against current solution either, which comes handy if you need default behaviour. For example, OkHttp does the same thing with its empty constructor.